src/router/validation/subscriptionValidator.ts

```diff
@@ -0,0 +1,135 @@
+
+export default class SubscriptionValidator implements Validator {
+    private ajv: Ajv.Ajv;
+
+    private validateJSON: Ajv.ValidateFunction;
+
+    private search: Search;
+
+    private allowListMap: { [key: string]: (string | RegExp)[] } = {};
+
+    private allowListLoaded = false;
+
+    private readonly getSubscriptionAllowEndpoint: GetAllowEndpoints;
+
+    private readonly enableMultiTenancy: boolean;
+
+    constructor(search: Search, getSubscriptionAllowEndpoint: GetAllowEndpoints, enableMultiTenancy: boolean) {
+        this.search = search;
+        this.getSubscriptionAllowEndpoint = getSubscriptionAllowEndpoint;
+        this.enableMultiTenancy = enableMultiTenancy;
+        this.ajv = new Ajv();
+        this.validateJSON = this.ajv.compile(subscriptionSchema);
+    }
+
+    async loadAllowList() {
+        const allowEndpoints: SubscriptionEndpoint[] = await this.getSubscriptionAllowEndpoint();
+        if (!this.enableMultiTenancy) {
+            this.allowListMap = {
+                [SINGLE_TENANT_ALLOW_LIST_KEY]: allowEndpoints.map(
+                    (allowEndpoint: SubscriptionEndpoint) => allowEndpoint.endpoint,
+                ),
+            };
+        } else {
+            const endpointsGroupByTenant: { [key: string]: SubscriptionEndpoint[] } = groupBy(
+                allowEndpoints,
+                (allowEndpoint: SubscriptionEndpoint) => allowEndpoint.tenantId,
+            );
+            Object.entries(endpointsGroupByTenant).forEach(([key, value]) => {
+                this.allowListMap[key] = value.map((v) => v.endpoint);
+            });
+        }
+        this.allowListLoaded = true;
+    }
+
+    async validate(resource: any, tenantId?: string): Promise<void> {
+        const resourcesToValidate: any[] = this.filterResourcesToValidate(resource);
+        if (isEmpty(resourcesToValidate)) {
+            return;
+        }
+        if (!this.allowListLoaded) {
+            await this.loadAllowList();
+        }
+        const allowList: (string | RegExp)[] = this.getAllowListForRequest(tenantId);
+
+        resourcesToValidate.forEach((res) => {
+            const result = this.validateJSON(res);
+            if (!result) {
+                throw new InvalidResourceError(
+                    `Subscription resource is not valid. Error was: ${this.ajv.errorsText(this.validateJSON.errors)}`,
+                );
+            }
+            if (!this.isEndpointAllowListed(allowList, res.channel.endpoint)) {
+                throw new InvalidResourceError(
+                    `Subscription resource is not valid. Endpoint ${res.channel.endpoint} is not allow listed.`,
+                );
+            }
+            this.search.validateSubscriptionSearchCriteria(res.criteria);
+        });
+    }
+
+    // eslint-disable-next-line class-methods-use-this
+    private isEndpointAllowListed(allowList: (string | RegExp)[], endpoint: string): boolean {
```

Comment:
Let's not disable `class-methods-use-this`. this method can be changed to a fuction outisde of the class. ([example](https://github.com/awslabs/fhir-works-on-aws-search-es/blob/9847e1415d5067496b391e733fdd365185a062f0/src/FHIRSearchParametersRegistry/index.ts#L67))

I know that our codebase disables `class-methods-use-this` a lot, but going forward we should avoid it.